src/servers/src/http/event.rs

```diff
@@ -0,0 +1,241 @@
+
+    if payload.is_empty() {
+        return Err(InvalidParameterSnafu {
+            reason: "pipeline is required in body",
+        }
+        .build());
+    }
+
+    let content_type = "yaml";
+    let result = handler
+        .insert_pipeline(&pipeline_name, content_type, &payload, query_ctx)
+        .await;
+
+    result.map(|_| "ok".to_string()).map_err(|e| {
+        error!(e; "failed to insert pipeline");
+        e
+    })
+}
+
+/// Transform NDJSON array into a single array
+fn transform_ndjson_array_factory(
+    ignore_error: bool,
+) -> impl FnMut(Result<Value>, std::result::Result<Value, serde_json::Error>) -> Result<Value> {
+    move |acc, item| {
+        acc.and_then(|acc| match acc {
+            Value::Array(mut acc_array) => {
+                if let Ok(item_value) = item {
+                    match item_value {
+                        Value::Array(item_array) => {
+                            acc_array.extend(item_array);
+                        }
+                        Value::Object(_) => {
+                            acc_array.push(item_value);
+                        }
+                        _ => {
+                            if !ignore_error {
+                                warn!("invalid item in array: {:?}", item_value);
+                                return Err(InvalidParameterSnafu {
+                                    reason: format!("invalid item:{} in array", item_value),
+                                }
+                                .build());
+                            }
+                        }
+                    }
+                    Ok(Value::Array(acc_array))
+                } else if !ignore_error {
+                    item.context(ParseJsonSnafu)
+                } else {
+                    warn!("invalid item in array: {:?}", item);
+                    Ok(Value::Array(acc_array))
+                }
+            }
+            _ => unreachable!("invalid acc: {:?}", acc),
+        })
+    }
+}
+
+#[axum_macros::debug_handler]
+pub async fn log_ingester(
+    State(handler): State<LogHandlerRef>,
+    Query(query_params): Query<LogIngesterQueryParams>,
+    Extension(query_ctx): Extension<QueryContextRef>,
+    TypedHeader(content_type): TypedHeader<ContentType>,
+    payload: String,
+) -> Result<HttpResponse> {
+    let pipeline_name = query_params.pipeline_name.context(InvalidParameterSnafu {
+        reason: "pipeline_name is required",
+    })?;
+    let table_name = query_params.table.context(InvalidParameterSnafu {
+        reason: "table is required",
+    })?;
+
+    let version = query_params.version;
+
+    let ignore_errors = query_params.ignore_errors.unwrap_or(false);
+
+    let m: mime::Mime = content_type.clone().into();
+    let value = match m.subtype() {
+        mime::JSON => Deserializer::from_str(&payload).into_iter::<Value>().fold(
+            Ok(Value::Array(Vec::with_capacity(100))),
+            transform_ndjson_array_factory(ignore_errors),
```

Comment:
do we treat JSON as ndjson? then we won't be able to deal with pure json?

Reply:
We can deal with pure json.
There is no difference between a normal json object and ndjson. It's mainly the json array that causes ambiguity. This function is designed to deal with some of the points that are easy to misuse.